src/lib.rs

```diff
@@ -712,40 +1006,125 @@ where
       &recursive_snark.l_w_secondary,
     )?;
 
+    // derandomize/unblind commitments
+    let (derandom_r_U_primary, derandom_r_W_primary) = recursive_snark
+      .r_U_primary
+      .derandomize_commits_and_witnesses(&pp.ck_primary, &recursive_snark.r_W_primary);
+    let (derandom_f_U_secondary, derandom_f_W_secondary) =
+      f_U_secondary.derandomize_commits_and_witnesses(&pp.ck_secondary, &f_W_secondary);
+
     // create SNARKs proving the knowledge of f_W_primary and f_W_secondary
     let (r_W_snark_primary, f_W_snark_secondary) = rayon::join(
       || {
         S1::prove(
           &pp.ck_primary,
           &pk.pk_primary,
           &pp.r1cs_shape_primary,
-          &recursive_snark.r_U_primary,
-          &recursive_snark.r_W_primary,
+          &derandom_r_U_primary,
+          &derandom_r_W_primary,
         )
       },
       || {
         S2::prove(
           &pp.ck_secondary,
           &pk.pk_secondary,
           &pp.r1cs_shape_secondary,
-          &f_U_secondary,
-          &f_W_secondary,
+          &derandom_f_U_secondary,
+          &derandom_f_W_secondary,
         )
       },
     );
 
     Ok(Self {
       r_U_primary: recursive_snark.r_U_primary.clone(),
+      ri_primary: recursive_snark.ri_primary,
       r_W_snark_primary: r_W_snark_primary?,
 
       r_U_secondary: recursive_snark.r_U_secondary.clone(),
+      ri_secondary: recursive_snark.ri_secondary,
       l_u_secondary: recursive_snark.l_u_secondary.clone(),
       nifs_secondary,
       f_W_snark_secondary: f_W_snark_secondary?,
 
       zn_primary: recursive_snark.zi_primary.clone(),
       zn_secondary: recursive_snark.zi_secondary.clone(),
 
+      primary_blind_r_W: recursive_snark.r_W_primary.r_W,
+      primary_blind_r_E: recursive_snark.r_W_primary.r_E,
+      secondary_blind_r_W: f_W_secondary.r_W,
+      secondary_blind_r_E: f_W_secondary.r_E,
+      random_layer: None,
+      _p: Default::default(),
+    })
+  }
+
+  /// Create a new `CompressedSNARK` (with randomizing layer)
+  fn prove_randomizing(
```

Comment:
I think given that adding zk is a negligible/tiny cost to the prover and the verifier and proof sizes, I think we should make it the default behavior. This will completely avoid a bunch of code. That is, we could just "upgrade" the existing CompressedSNARK type to provide the zero-knowledge property rather than introducing new methods that one will have to use.

Reply:
What do you think @jkwoods?

Reply:
This seems good to me! I'll edit.